src/schema.cpp

```diff
@@ -82,29 +82,35 @@ Schema::const_iterator Schema::find(ObjectSchema const& object) const noexcept
 namespace {
 
 struct CheckObjectPath {
-    const ObjectSchema& object;
-    std::string path;
+    const ObjectSchema& object; // the schema to check
+    std::string path; // a printable path for error messaging
+    std::unordered_set<std::string> visited; // the list of object types encountered along this path so far
```

Comment:
I know you're just copying this from elsewhere in this function, but how many items do we expect visited to contain? std::unordered_set might be overkill here.

Reply:
There will only be a handful of items yes. I think the data structure does does convey the intent though; ie a bucket of objects that we've already seen on the path in no particular order. I don't think this is worth measuring performance for different structures because schema checking is not a hot path.

Reply:
That's fine if this is a cold path. I agree the data structure matches the intent. I at least wanted to ask if we need to construct a hash table to store ~10-100 short strings when a vector and `std::find()` would work just as well.